### PR TITLE
Optimize semi in join with aggregate window functions

### PR DESCRIPTION
**Pull request by bhuvnesh2703**
Decorrelate queries with Aggregate Window Func. 
If the correlated subquery has aggregate window functions we can
pull up the quals and aggregate function as a the condition of the
join between outer and inner query

Previous plan
```
  create table B(i integer, j integer);
  insert into B values(1,1);
  insert into B values(2,2);
  insert into B values(2,2);
  insert into B values(3,3);
  analyze B;
  create table C(i integer, j integer);
  insert into C values(1,1);
  insert into C values(1,1);
  insert into C values(2,2);
  analyze C;
bhuvneshchaudhary=# explain select C.i from C where C.i in (select avg(i) over (partition by i) from B  where C.i = B.i) order by C.j;
                                                     QUERY PLAN
--------------------------------------------------------------------------------------------------------------------
 Result  (cost=0.00..1324032.56 rows=1 width=4)
   Filter: (subplan)
   ->  Sort  (cost=0.00..431.00 rows=1 width=4)
         Sort Key: c.j
         ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=8)
               ->  Table Scan on c  (cost=0.00..431.00 rows=1 width=8)
   SubPlan 1
     ->  Result  (cost=0.00..431.00 rows=1 width=8)
           ->  Window  (cost=0.00..431.00 rows=1 width=8)
                 Partition By: b.i
                 ->  Sort  (cost=0.00..431.00 rows=1 width=4)
                       Sort Key: b.i
                       ->  Result  (cost=0.00..431.00 rows=1 width=4)
                             Filter: $1 = b.i
                             ->  Materialize  (cost=0.00..431.00 rows=1 width=4)
                                   ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..431.00 rows=1 width=4)
                                         ->  Table Scan on b  (cost=0.00..431.00 rows=1 width=4)
 Optimizer status: PQO version 2.45.0
```
After pulling up clause;
```
                                          QUERY PLAN
-----------------------------------------------------------------------------------------------
 Result  (cost=0.00..862.00 rows=1 width=4)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..862.00 rows=1 width=4)
         Merge Key: c.j
         ->  Sort  (cost=0.00..862.00 rows=1 width=8)
               Sort Key: c.j
               ->  Hash EXISTS Join  (cost=0.00..862.00 rows=1 width=8)
                     Hash Cond: c.i = b.i AND c.i::numeric = (avg(b.i) OVER (?))
                     ->  Table Scan on c  (cost=0.00..431.00 rows=1 width=8)
                     ->  Hash  (cost=431.00..431.00 rows=1 width=12)
                           ->  Window  (cost=0.00..431.00 rows=1 width=12)
                                 Partition By: b.i
                                 ->  Sort  (cost=0.00..431.00 rows=1 width=4)
                                       Sort Key: b.i
                                       ->  Table Scan on b  (cost=0.00..431.00 rows=1 width=4)
 Optimizer status: PQO version 2.45.0
```
Results:
```
bhuvneshchaudhary=# set optimizer=on;
SET
bhuvneshchaudhary=# select C.i from C where C.i in (select avg(i) over (partition by i) from B  where C.i = B.i) order by C.j;
 i
---
 1
 1
 2
(3 rows)
```

Data in tables
```
bhuvneshchaudhary=# select * from C;
 i | j
---+---
 1 | 1
 1 | 1
 2 | 2
(4 rows)

bhuvneshchaudhary=# select * from B;
 i | j
---+---
 1 | 1
 2 | 2
 2 | 2
 3 | 3
(4 rows)
```

### CODE REVIEW

**Comment by hsyuan** [comment]
Nice work! 
A few suggestions: 
1. Can you provide DDL and data generations in your PR description, instead of just providing select result in the table (which is GSS style)? 
1. Can you remove planner results here? This PR has nothing to do with planner. It will be confusing.

File: data/dxl/minidump/CorrelatedSubqueryWithAggWindowFunc.mdp [hunk 1]
```diff
@@ -0,0 +1,591 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (hsyuan) [inline data/dxl/minidump/CorrelatedSubqueryWithAggWindowFunc.mdp]: DDL and query?

File: libgpopt/src/xforms/CDecorrelator.cpp [hunk 2]
```diff
@@ -680,11 +700,20 @@ CDecorrelator::FProcessProject
...
+		// fail decorrelation in the following two cases;
+		// 1. if the LogicalSequenceProject node has local outer references in order by or partition by or window frame
+		// of a window function
+		// ex: select C.j from C where C.i in (select rank() over (order by C.i) from B where B.i=C.i);
+		// 2. if the relational child of LogicalSequenceProject node does not have any aggregate window function
+
+		// if the project list contains aggregrate on window function, then
+		// we can decorrelate it as the aggregate is performed over a column or count(*).
+		// The IN condition will be translated to a join instead of a correlated plan.
+		// ex: select C.j from C where C.i in (select avg(i) over (partition by B.i) from B where B.i=C.i);
```
Comment (hsyuan) [inline libgpopt/src/xforms/CDecorrelator.cpp]: What if the window definition is `(partition by B.i order by B.j)`? Can we still decorrelate it?

Comment (hsyuan) [reply libgpopt/src/xforms/CDecorrelator.cpp]: hmm, yeah, I think we can.

File: data/dxl/minidump/CorrelatedSubqueryWithAggWindowFunc.mdp (hunk 1 above)
Comment (bhuvnesh2703) [reply data/dxl/minidump/CorrelatedSubqueryWithAggWindowFunc.mdp]: added

File: libgpopt/src/xforms/CDecorrelator.cpp (hunk 2 above)
Comment (bhuvnesh2703) [reply libgpopt/src/xforms/CDecorrelator.cpp]: yes, we can. Order by has not impacts on the results unless there is limit. And if there is a limit we donot decorrelate it.

**Comment by bhuvnesh2703** [comment]
Thanks @hsyuan included the ddl, data and removed planner results.

File: server/src/unittest/gpopt/minidump/CWindowTest.cpp [hunk 3]
```diff
@@ -34,6 +34,7 @@ CWindowTest::EresUnittest()
 			"../data/dxl/minidump/Preds-Over-WinFunc4.mdp",
 			"../data/dxl/minidump/Preds-Over-WinFunc5.mdp",
 			"../data/dxl/minidump/SemiJoinWithWindowsFuncInSubquery.mdp",
+			"../data/dxl/minidump/CorrelatedSubqueryWithAggWindowFunc.mdp",
```
Comment (oarap) [inline server/src/unittest/gpopt/minidump/CWindowTest.cpp]: This doesn't feel like the right place for this test. `CSubqueryHandlerTest` or `CDecorrelatorTest` or even `CICGTest` feels better for me.

Comment (bhuvnesh2703) [reply server/src/unittest/gpopt/minidump/CWindowTest.cpp]: sure will move this.

**Pull request merged**